src/charm.py

```diff
@@ -1,104 +1,87 @@
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
-                }
-            },
-        }
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ActiveStatus()
-
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
-
-        Learn more about config at https://juju.is/docs/sdk/config
-        """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
-
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
-
-        Learn more about actions at https://juju.is/docs/sdk/actions
-        """
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
-        else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+
+        self.name = "mysqlrouter"
+        self.framework.observe(self.on.install, self._on_install)
+
+    def _on_install(self, _) -> None:
+        """Install the packages."""
+        self.unit.status = MaintenanceStatus("installing packages")
+
+        self._install_apt_packages([MYSQL_ROUTER])
+        self._install_snap_packages([MYSQL_SHELL])
+        self.unit.status = WaitingStatus("waiting for database relation")
+
+    def _install_apt_packages(self, packages: List[str]) -> None:
+        """Install apt packages."""
+        try:
+            logger.debug("Updating apt cache")
+            apt.update()
+        except subprocess.CalledProcessError as e:
+            logger.exception("Failed to update apt cache", exc_info=e)
+            self.unit.status = BlockedStatus("failed to update apt cache")
+            raise e
```

Comment:
im not sure what will happen if we `raise e` (or even `raise` - to preserve the original stack trace) will do to the unit. would it enter an `errored` state with no chance of recovery?

Reply:
It will, and I'm not comfortable with it. Open discussion that I will bring to the table on charm open hours.